Extend reroute with an option to force assign stale primary shard copies

ywelsch opened the pull request:
Relates to #14739.

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,297 @@
+
+            String currentFieldName = null;
+            XContentParser.Token token;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```
Should this use the `ObjectParser`? I think everything is going that way.

ywelsch replied:
I essentially copy pasted this stuff from the `AllocateAllocationCommand`. You are right though, I should transform it to use `ObjectParser`.

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,297 @@
+            } else {
+                builder.startObject(objectName);
+            }
+            builder.field("index", command.shardId().index().name());
```
Other places we use a `Fields` inner class for the field names. Not here? I'm not really 100% sure why we don't just use strings for the field names so I can't speak with confidence about this.

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,297 @@
+        // check if we have a primary unassigned shard
+        ShardRouting shardRouting = null;
+        for (ShardRouting routing : routingNodes.unassigned()) {
+            if (routing.shardId().equals(shardId)) {
```
Smoosh these into one if statement? Its not a requirement, I'm more curious why you didn't do it.

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,297 @@
+        }
+
+        Decision decision = allocation.deciders().canAllocate(shardRouting, routingNode, allocation);
+        if (ignoreDeciders == false && decision.type() == Decision.Type.NO) {
```
Why even ask if `ignoreDeciders == true`?

ywelsch replied:
The idea behind that was so that we could see in the RerouteExplanation if it was Yes, Forced Yes or Throttle. It's probably simpler to just always provide the decision  (YES, "override deciders") in case of explicit override.

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,297 @@
+            break;
+        }
+        if (ignoreDeciders && decision.type() == Decision.Type.NO) {
+            decision = allocation.decision(Decision.YES, RECOVER_PRIMARY_ALLOCATION_COMMAND, "override deciders");
```
This seems mighty dangerous but I get it. I suspect things will start relocating as soon as you do this too.

ywelsch replied:
It is dangerous, yes. This flag should only be used with extreme care and needs to be documented as such. The reason it's there is to allow manual primary allocation to be as powerful as for the automatic case. In the automatic case, we force allocation even if all nodes say no (see PrimaryShardAllocator: https://github.com/elastic/elasticsearch/blob/c50b22f95f48119213156bdb634925e1be91b794/core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java#L140 ). As you noted, this can trigger relocation of other shards.

bleskes commented:
@ywelsch and I discussed this yesterday. Here is the gist of the discussion:

1) Instead of fetching the store data in the command, adapt unassigned info so that the node initializing the shard will expect local data and fail if it's not there.
2) Remove the ignore allocation decider param and always assign if the user asked for it (in the case of the primary assignment)
3) Try to fold the new AllocatePrimaryAllocationCommand in the current AllocateAllocationCommand.

Here's a suggest for how the rest layer will look like for that:

```
POST _cluster/reroute
{
  "commands": [
    {
      "allocate": {
        "index": "index",
        "shard": 0,
        "node": "node",
        "allow_primary": "never|existing_only|force_empty"
      }
    }
  ]
```

Note the `allow_primary` option. The default, `never` refuses to allocate a primary. `existing_only` allocates it but must reuse an existing copy. `force_empty` does what true does now - i.e., force the creation of an empty primary. Also the last two should imply ignoring the allocation deciders.

@ywelsch thoughts?

ywelsch commented:
Before this PR, users only had the possibility to either allocate a replica or an empty primary. The flag "allow_primary" was misleading in that context. I would very much like to get rid of this flag altogether and introduce something new. What about introducing a new "mode" field:

"mode": "replica" // default, only allocates unassigned replica shard (fails if unassigned primary exists)
"mode": "primary" // only allocates if primary unassigned and if data available on node
"mode": "empty_primary" // only allocates if primary unassigned, creates empty shard

"mode": "replica" is equivalent to the old "allowPrimary" : "false"
"mode": "empty_primary" is equivalent to the old "allowPrimary" : "true"

ywelsch commented:
Also, relates to #14352

bleskes commented:
I think mode is a bit ambiguous as a flag. I do understand the concerns that `allow_primary` will now do something different than the allow primary in the cancel command.  As far as the values. I like "existing_only" better than just "primary" as it better communicates the restriction (we must use local data). Also I like "force_empty" better than "empty_primary" because it communicates the implications better and the severity better. It also better describes what we do today - always create an empty shard, even if there is local data. That last one is of course subject to change (but that's a bigger discussion)

ywelsch commented:
>  As far as the values. I like "existing_only" better than just "primary" as it better communicates the restriction (we must use local data). Also I like "force_empty" better than "empty_primary" because it communicates the implications better and the severity better. It also better describes what we do today - always create an empty shard, even if there is local data.

```
$ grep -o "better" | wc -l
6
```

sorry ;-)

clintongormley commented:
We have a stalled issue (https://github.com/elastic/elasticsearch/issues/14352) about choosing a better name for `allow_primary` which properly indicates the fact that you can lose your data.  The issue is stalled because it is difficult to come up with the right name.

Instead, what about changing the action name from `allocate` to:

* `allocate_replica`
* `allocate_empty_primary` 
* and for this PR: `allocate_stale_primary`

@s1monw is concerned about exposing these last two actions via the API and would prefer to provide a command line tool (perhaps as a plugin) to do this job instead.  The idea is to make it harder for users to make a change that will delete data, and will hopefully force them to read about the consequences of their action.

While I understand this sentiment, I'm not convinced that a command line tool is the way to go here.  ES is API driven so this would be quite a change from what we do today, and there are probably many users who don't have command line access to their hosted clusters, and who need to fix a problem quickly without having to wait for support to respond to their tickets.

Instead, I'd propose adding an `accept_data_loss` flag to these actions which must be set to true for the action to be taken.  I think this would be enough to warn the user that they're doing something dangerous.

bleskes commented:
+1. My only suggestion would be to use 'force_empty_primary' and 'force_stale_primary' as command names, to make it even clearer that (all) shard content will be lost.


On 9 jan. 2016 12:13 PM +0100, Clinton Gormley<notifications@github.com>, wrote:
>  
> We have a stalled issue (#14352(https://github.com/elastic/elasticsearch/issues/14352)) about choosing a better name forallow_primarywhich properly indicates the fact that you can lose your data. The issue is stalled because it is difficult to come up with the right name.
>  
>  
> Instead, what about changing the action name fromallocateto:
>  
> allocate_replica
> allocate_empty_primary
> and for this PR:allocate_stale_primary
>  
>  
> @s1monw(https://github.com/s1monw)is concerned about exposing these last two actions via the API and would prefer to provide a command line tool (perhaps as a plugin) to do this job instead. The idea is to make it harder for users to make a change that will delete data, and will hopefully force them to read about the consequences of their action.
>  
>  
> While I understand this sentiment, I'm not convinced that a command line tool is the way to go here. ES is API driven so this would be quite a change from what we do today, and there are probably many users who don't have command line access to their hosted clusters, and who need to fix a problem quickly without having to wait for support to respond to their tickets.
>  
>  
> Instead, I'd propose adding anaccept_data_lossflag to these actions which must be set to true for the action to be taken. I think this would be enough to warn the user that they're doing something dangerous.
>  
>  
> —
> Reply to this email directly orview it on GitHub(https://github.com/elastic/elasticsearch/pull/15708#issuecomment-170226197).
>  
>  
>

ywelsch commented:
@bleskes @clintongormley I pushed a new set of changes reflecting the naming suggested by @clintongormley. This new changeset also contains documentation for the new commands. Two open points remain for me:

-  I am still not fully convinced about the command naming. I like @clintongormley's suggestion of command names `allocate_replica`, `allocate_empty_primary` and `allocate_stale_primary`. The `accept_data_loss` flag, however, feels unnatural to me in the rest api and even weirder in the Java API. I am not completely against a safeguard, but would prefer to have it in another form. As an inspiration, we might look at the way it is done for wildcard indices deletion ([`action.destructive_requires_name`](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-delete-index.html)).
- To keep the commands symmetric, we should also split the `cancel` command (which currently has an `allowPrimary` flag) into two commands `cancel_replica` and `cancel_primary`.

@bleskes What I don't like about `force_empty_primary` is that the action verb `allocate` is lost.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java:
```diff
@@ -0,0 +1,237 @@
+        out.writeString(node);
+    }
+
+    public static abstract class Factory<T extends AbstractAllocateAllocationCommand> implements AllocationCommand.Factory<T> {
```
Since this is a major rewrite anyway, can we try to move from custom factories  to `NamedWriteableRegistry` ?

bleskes replied:
I see now that MoveAllocationCommand is not touched by the PR. I think moving to NamedWriteableRegistry is a good idea, but I'm fine with putting it out of scope for this PR

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java:
```diff
@@ -0,0 +1,237 @@
+    /**
+     * Gets reroute explanation for a routing node that cannot be found
+     */
+    protected RerouteExplanation getMissingRoutingNodeExplanation(RoutingAllocation allocation, boolean explain, DiscoveryNode discoNode) {
```
can call this `explainOrThrowMissingRoutingNode` ? the docs can read something like "a utility method to handle the case where a disco node can not be found in the routing table. Typically this would mean it's not a data node"

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java:
```diff
@@ -0,0 +1,237 @@
+    /**
+     * Gets the first unassigned primary shard routing for the current shard id. Returns null if it can't find any.
+     */
+    protected ShardRouting getUnassignedPrimaryShardRouting(RoutingNodes routingNodes) {
```
Callers of this method can just do `allocation.routingTable().shardRoutingTable(shardId).primaryShard()` these days (if we add an overload for shardRoutingTable which takes a shardId). I don't think it's worth having this utility method. (I know it existed before - we have progressed since it was written :))

ywelsch replied:
This does not work unfortunately. The issue is that the ShardRouting instance in the RoutingTable is a different instance than the one in RoutingNodes. `initializeUnassignedShard` which uses reference equality does not match the proper shard then.

bleskes replied:
hmm. That smells. How about using `ShardRouting#equalsIgnoringMetaData` there?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java:
```diff
@@ -0,0 +1,237 @@
+            }
+            it.initialize(routingNode.nodeId(), unassigned.version(), allocation.clusterInfo().getShardSize(unassigned, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE));
+            break;
+        }
```
can we assert we found the requested shard in the loop?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java:
```diff
@@ -0,0 +1,237 @@
+    /**
+     * Creates negative reroute explanation based on provided text
+     */
+    protected RerouteExplanation getNegativeRerouteExplanation(boolean explain, RoutingAllocation allocation, String text) {
```
same - please name it something like `explainOrThrowRejectedCommand`

bleskes replied:
another nit: call the `text` param, reason?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java:
```diff
@@ -0,0 +1,237 @@
+     */
+    protected RerouteExplanation getNegativeRerouteExplanation(boolean explain, RoutingAllocation allocation, String text) {
+        if (explain) {
+            return new RerouteExplanation(this, allocation.decision(Decision.NO, name() + "_allocation_command", text));
```
maybe just use `name() + " (allocation command)"` ?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,112 @@
+
+        ShardRouting shardRouting = getUnassignedPrimaryShardRouting(routingNodes);
+        if (shardRouting == null) {
+            return getNegativeRerouteExplanation(explain, allocation, "failed to find primary " + shardId + " on the list of unassigned shards");
```
maybe the reason can simply be "primary " + shardId + " is already assigned".

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,112 @@
+
+        if (acceptDataLoss == false) {
+            return getNegativeRerouteExplanation(explain, allocation,
+                "trying to allocate an empty primary shard " + shardId + " that entails possible data loss, needs to be explicitly enabled");
```
Can this read `allocating an empty primary for " + shardId + " can result in data loss. please confirm by setting the accept_data_loss parameter to true.` ?

bleskes replied:
also, I don't think we need to do this if unassignedInfo.getReason == INDEX_CREATED?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,88 @@
+/**
+ * Abstract base class for allocating an unassigned primary shard to a node
+ */
+public abstract class AllocatePrimaryAllocationCommand extends AbstractAllocateAllocationCommand {
```
can we maybe call this BasePrimaryAllocationCommand? got confused while reading the extends clause somewhere else thinking this is a real command until I saw the abstract keyword here.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -0,0 +1,113 @@
+
+        ShardRouting shardRouting = getUnassignedReplicaShardRouting(routingNodes);
+        if (shardRouting == null) {
+            return getNegativeRerouteExplanation(explain, allocation, "failed to find " + shardId + " on the list of unassigned shards");
```
maybe "all copies of " + shardId +" are already assigned. use the move allocation command instead".

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java:
```diff
@@ -0,0 +1,112 @@
+
+        ShardRouting shardRouting = getUnassignedPrimaryShardRouting(routingNodes);
+        if (shardRouting == null) {
+            return getNegativeRerouteExplanation(explain, allocation, "failed to find primary " + shardId + " on the list of unassigned shards");
```
see text from other suggestion for empty primary allocation

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java:
```diff
@@ -41,7 +53,7 @@
 public class PrimaryAllocationIT extends ESIntegTestCase {
 
-    public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception {
+    private void createStaleReplicaScenario() throws Exception {
```
nice!

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java:
```diff
@@ -103,6 +119,48 @@ public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception
+            .put("index.routing.allocation.exclude._name", node)
+            .put("index.number_of_shards", 1).put("index.number_of_replicas", 0)).get();
+
+        ClusterHealthResponse health = client().admin().cluster()
```
create does an inline reroute, so you can check directly here that no shard is assigned. No need for timeouts

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
+        try {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateEmptyPrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
+        } catch (IllegalArgumentException e) {
```
can we check the message here?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
+        logger.info("--> allocating empty primary with acceptDataLoss flag set to false");
+        try {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateEmptyPrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
```
can you add a message to this fail?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
+        logger.info("--> allocating stale primary with acceptDataLoss flag set to false");
+        try {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateStalePrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
```
same as above

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateStalePrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
+        } catch (IllegalArgumentException e) {
+        }
```
same as above

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -210,7 +235,7 @@ public void testCancelCommand() {
         assertThat(clusterState.getRoutingNodes().shardsWithState(INITIALIZING).size(), equalTo(0));
 
         logger.info("--> allocating with primary flag set to true");
```
stale logging message - use allow data loss set to true

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java:
```diff
@@ -103,6 +119,48 @@ public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception
+        client().admin().cluster().prepareReroute().add(new AllocateEmptyPrimaryAllocationCommand(new ShardId("test", 0), node, true)).get();
+        ensureGreen("test");
+    }
+
```
can we add a variant that tests that assigning a stale primary to a node that doesn't have the data fails to initialize the shard?

bleskes commented on docs/reference/cluster/reroute.asciidoc:
```diff
@@ -64,14 +64,41 @@ The commands supported are:
+`allocate_stale_primary`::
+    Allocate an unassigned stale primary shard to a node. Accepts the
+    `index` and `shard` for index name and shard number, and `node` to
+    allocate the shard to. Using this command leads to a loss of data
```
may lead to data loss..

bleskes commented on docs/reference/cluster/reroute.asciidoc:
```diff
@@ -64,14 +64,41 @@ The commands supported are:
+are available:
+
+`allocate_stale_primary`::
+    Allocate an unassigned stale primary shard to a node. Accepts the
```
Allocates a primary shard to a node that holds a stale copy.

bleskes commented on docs/reference/cluster/reroute.asciidoc:
```diff
@@ -64,14 +64,41 @@ The commands supported are:
+`allocate_empty_primary`::
+    Allocate an unassigned empty primary shard to a node. Accepts the
+    `index` and `shard` for index name and shard number, and `node` to
+    allocate the shard to. Using this command leads to a complete loss
```
Using this command leads to a complete loss of all data that was indexed into this shard, if it was previously started.

bleskes commented:
Thanks @ywelsch . I really like how it turned out. Left some minor comments here and there.

ywelsch commented:
pushed another round of changes addressing the comments. Please have another look @bleskes. (The `NamedWriteableRegistry` refactoring will be for another time)

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java:
```diff
@@ -103,6 +116,63 @@ public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception
     }
 
+    public void testFailedAllocationOfStalePrimaryToDataNodeWithNoData() throws Exception {
+        createStaleReplicaScenario();
```
I think this can be made simpler by just creating a one node cluster, create an index with no replicas, add a node to the cluster. Shutdown the first node and force assign. Makes sense?

ywelsch replied:
ok

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -107,9 +121,10 @@ public void testAllocateCommand() {
                 .build();
         RoutingTable routingTable = RoutingTable.builder()
-                .addAsNew(metaData.index("test"))
+                .addAsRecovery(metaData.index("test"))
```
why was this needed?

ywelsch replied:
because later we test that allocating empty primary with accept_data_loss flag set to false fails.

bleskes replied:
kk. cool. Can you add a comment?

> On 19 Jan 2016, at 11:42, Yannick Welsch <notifications@github.com> wrote:
> 
> In core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
> 
> > @@ -107,9 +121,10 @@ public void testAllocateCommand() {
> >                  .put(IndexMetaData.builder("test").settings(settings(Version.CURRENT)).numberOfShards(1).numberOfReplicas(1))
> >                  .build();
> >          RoutingTable routingTable = RoutingTable.builder()
> > -                .addAsNew(metaData.index("test"))
> > +                .addAsRecovery(metaData.index("test"))
> 
> because later we test that allocating empty primary with accept_data_loss flag set to false fails.
> 
> —
> Reply to this email directly or view it on GitHub.
>

ywelsch replied:
will do

bleskes commented:
LGTM. Made one suggestion for simplifying a test (and asked a question). Feel free to adopt the suggestion (or not) and push. Thanks for all the iterations, @ywelsch .

ywelsch commented:
Thanks for the thorough review @bleskes !
